MediaGallerySynchronization/Model/AssetsBatchGenerator.php

```diff
@@ -0,0 +1,154 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\MediaGallerySynchronization\Model;
+
+use Magento\Framework\Filesystem;
+use Magento\Framework\App\Filesystem\DirectoryList;
+use Magento\Framework\Filesystem\Directory\Read;
+use Magento\MediaGalleryApi\Api\IsPathBlacklistedInterface;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Batch generator for filesystem files, get files by providet size of batch
+ */
+class AssetsBatchGenerator
+{
+    private const IMAGE_FILE_NAME_PATTERN = '#\.(jpg|jpeg|gif|png)$# i';
+    
+    /**
+     * @var GetAssetsIterator
+     */
+    private $getAssetsIterator;
+
+    /**
+     * @var Filesystem
+     */
+    private $filesystem;
+
+    /**
+     * @var Read
+     */
+    private $mediaDirectory;
+
+    /**
+     * @var IsPathBlacklistedInterface
+     */
+    private $isPathBlacklisted;
+
+    /**
+     * @var array
+     */
+    private $assetsPaths;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $log;
+
+    /**
+     * @var ResolveNonExistedAssets
+     */
+    private $resolveNonExistedAssets;
+    
+    /**
+     * @param LoggerInterface $log
+     * @param ResolveNonExistedAssets $resolveNonExistedAssets
+     * @param IsPathBlacklistedInterface $isPathBlacklisted
+     * @param Filesystem $filesystem
+     * @param GetAssetsIterator $assetsIterator
+     */
+    public function __construct(
+        LoggerInterface $log,
+        ResolveNonExistedAssets $resolveNonExistedAssets,
+        IsPathBlacklistedInterface $isPathBlacklisted,
+        Filesystem $filesystem,
+        GetAssetsIterator $assetsIterator
+    ) {
+        $this->log = $log;
+        $this->resolveNonExistedAssets = $resolveNonExistedAssets;
+        $this->isPathBlacklisted = $isPathBlacklisted;
+        $this->getAssetsIterator = $assetsIterator;
+        $this->filesystem = $filesystem;
+        $this->assetsPaths = [];
+    }
+
+    /**
+     * Return files from files system by provided size of batch
+     *
+     * @param int $size
+     */
+    public function getItems(int $size): \Generator
+    {
+        $i = 0;
+        $batch = [];
+
+        /** @var \SplFileInfo $file */
+        foreach ($this->getAssetsIterator->execute($this->getMediaDirectory()->getAbsolutePath()) as $file) {
+            if (!$this->isApplicable($file->getPathName())) {
+                continue;
+            }
+            $this->assetsPaths[] = $this->getRelativePath($file->getPathName(), $file->getFileName());
+
+            $batch[] = $file;
+            if (++$i == $size) {
+                yield $batch;
+                $i = 0;
+                $batch = [];
+            }
+        }
+        if (count($batch) > 0) {
+            yield $batch;
+        }
+        
+        $this->resolveNonExistedAssets->execute($this->assetsPaths);
```

Comment:
I'd rather not include this operation in the batch generator. Could we extract the backward synchronization and perform it in batches as well?

Reply:
ok